### ISSUE TITLE
v6 validation: "constant"

### ISSUE THREAD

**Issue by handrews** [json-schema-org/json-schema-spec#58]
Originally written by @geraintluff at https://github.com/json-schema/json-schema/wiki/constant-(v5-proposal)
### Proposed keywords
- `constant`
### Purpose

For ordinary use, this would be equivalent to a single-valued `enum`, simply tidier.

The only real difference comes in its behaviour with `$data`-substitution.  `$data`-subsitution would be allowed for this keyword, which means that this keyword is _not_ capable of specifying literal values of the form: `{"$data":...}`, because these would be interpreted as `$data`-substitutions.

However, literal values of this form can still be specified using `enum`, so there is no loss of functionality.
### Values

The value of this keyword would be any value - however, it would be subject to `$data`-substitution.
### Validation

Instances are only valid if they are exactly equal to the value of this keyword.
#### Simple constant

``` json
{
    "type": "object",
    "properties": {
        "five": {
            "constant": 5
        }
    }
}
```

Valid: `{}`, `{"five": 5}`
Invalid: `{"five": 0}`, `{"five": "5"}`
#### Using `$data` to specify equality

``` json
{
    "type": "object",
    "properties": {
        "a": {"type": "string"},
        "b": {
            "constant": {"$data": "1/a"}
        }
    },
    "required": ["a", "b"]
}
```

Valid: `{"a": "foo", "b": "foo"}`, `{"a": "bar", "b": "bar"}`
Invalid: `{"a": "foo", "b": "bar"}`
#### Similarity to `enum`

Unless `$data` is being used, the same effect can be obtained using fewer actual characters:
- `{"constant":"whatever"}` - 23 characters
- `{"enum":["whatever"]}` - 21 characters

However, when used in combination with `$data`, it opens up possibilities that are not otherwise available.

**Comment by handrews** [comment]
I personally find `constant` to be useful independent of `$data` as it more clearly expresses the intent of the schema author to require a fixed value, and not an enum that (at the moment) happens to only have one entry.

**Comment by HotelDon** [comment]
Personally, I get the most use out of constant when combining it with $data, but you bring up an excellent point I hadn't considered with regards to ambiguity when you use enum. While I'm still pushing for $data, I think constant could stand on its own if it really needed to.

**Comment by sceutre** [comment]
Maybe "const" instead of "constant".  Seems more natural/widely spread as a keyword, and trounces the "enum" form in character count ;-)

**Comment by epoberezkin** [comment]
It is useful, particularly in combination with $data keyword - it allows to validate joins in objects from ORM systems.

**Comment by fny** [comment]
If you're really concerned about character count, you could always use `const`:

    {"constant":"whatever"} - 23 characters
    {"enum":["whatever"]} - 21 characters
    {"const":"whatever"} - 20 characters

**Comment by awwright** [comment]
Landed in #139 and 6873d6672d4300e04a172e8c389d5bd6918b7d36 -- does this version look appealing to everyone?

**Comment by handrews** [comment]
It's been over a week and no one has objected to the PR so I'm closing this as fixed.  Obviously can be re-opened if someone gets back from vacation or something and wants to raise it.
